Codeceptjs tests using Puppeteer.

pmk65 opened the pull request:
Codeceptjs tests using Puppeteer. Tests updated to work with CodeceptJS/Puppeteer

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks backward-compatibility?    | ❌(See comments)
| Tests pass?   | ❌ (See comments)
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any
| Updated README/docs?   | ❌
| Added CHANGELOG entry?   | ❌

* **codecept.conf.js** - Config file for CodeceptJS with Puppeteer
* **tests/codeceptjs/codecept.json** - Config file for CodeceptJS with WebDriver
* **config/puppeteer.helpers.js** - Custom helper functions (Used in both of the configs above)

There's some issues, but I don't know if it's related to the CodeceptJS/Puppeteer setup or my retarded Windows 10 Home Edition. 🗑 

### Issues
* If I turn off showing the browser (show: false), the tests fail.
* The test reports that popups are not closed. (I have updated the tests with `I.amCancellingPopups()` and `I.amAcceptingPopups()` which have solved most of the problems)
* I tried installing the `puppeteer-firefox` plugin, but failed to get it to work.
* `npm run ccserver` is Windows specific. (Need to find a better way to do this.)

### Changes
* I have updated the tests to work with Puppeteer as there where some problems with `I.pressKey()`. (example: `I.pressKey('__YELLOW__');` which is invalid as there's no key named `__YELLOW__`)
* Now only 3 tests fail (2 of them are @optional)
* Full test incl. @optional took 3:25 minutes.
* Uses [Mochawesome](https://www.npmjs.com/package/mochawesome) to generate nice HTML reports
![image](https://user-images.githubusercontent.com/478711/64111346-bce4d000-cd84-11e9-8160-4d9cdf935ad1.png)

pmk65 commented:
Seems like there's a difference between Docker and Puppeteer based tests. As `I.amCancellingPopups()` and `I.amAcceptingPopups()` are not recognized as functions in Docker. Also the Boolean checks for attributes is different. Docker wants **'true'**, but Puppeteer wants **true** (no quotes)
Se we might need to have a separate `tests/codeceptjs/` folder for each helper.

sirockin commented:
@pmk65 

> Seems like there's a difference between Docker and Puppeteer based tests. Se we might need to have a separate tests/codeceptjs/ folder for each helper.

I am firmly against maintaing two separate folders of tests - that will be a maintenance nightmare. Let's try to find an alternative.

pmk65 commented:
> 
> 
> @pmk65
> 
> > Seems like there's a difference between Docker and Puppeteer based tests. Se we might need to have a separate tests/codeceptjs/ folder for each helper.
> 
> I am firmly against maintaing two separate folders of tests - that will be a maintenance nightmare. Let's try to find an alternative.

I agree.
The quoted return value difference is easy to solve with some typecasting. And the `I.pressKey()` can be changed to `I.fillField()` in most cases, which works for both helpers.
The `I.cancelPopup()` and `I.acceptPopup()` should work without the additional `I.amCancellingPopups()` and `I.amAcceptingPopups()` function. It might just be a `wait` problem, as Puppeteer testing is REALLY fast compared to the Docker testing. (Or my retarded system.)

The output says:
> Popup already exists and was not closed. Popups must always be closed by calling either I.acceptPopup() or I.cancelPopup()

But the popups does close and some tests pass without having to add `I.amCancellingPopups()` and `I.amAcceptingPopups()`

pmk65 commented:
I have added some custom helpers. These works perfectly with Puppeteer.
But since I can't run Docker on my Win10 Home, I'm not able to test them with Docker/WebDriver. (So I could use some help here.)

sirockin commented:
Hi @pmk65. Thanks again for again for working on this and sorry not to have had a chance to review yet - I'm up against a bit of a deadline right now but will do my best to have a look later today.

pmk65 commented:
I have modified the tests and added custom helper functions so the same tests can run using either Puppeteer or WebDriver.
Currently there's a single error in the WebDriver test. (string) and one in Puppeteer (array).

schmunk42 commented:
I appreciate the work on tests in general, but before merging this, we should make sure that we do not introduce too much logic into the testing scenarios.

Also test should pass for Selenium and Pupeteer.

@pmk65 What exactly does Pupeteer? Does it control your local browser?

pmk65 commented:
> 
> 
> I appreciate the work on tests in general, but before merging this, we should make sure that we do not introduce too much logic into the testing scenarios.

There's really no extra logic added, except for custom helper functions to make the Puppeteer functions `I.amCancellingPopups` and `I.amAcceptingPopups` accepted by WebDriver (The helper used in the Selenium/Travis tests) 
Using custom helper functions we can create our own "I.blahblah" functions for use in the tests.

Other than that, I just changed invalid use of `I.pressKey` to `I.fillField` and changed checking for disabled/readonly to a more uniform method. (The method used before would return "", "false" or "true" in WebDriver and in Puppeteer null, false, true)


> Also test should pass for Selenium and Pupeteer.
> 
They do now, except for 2 tests. I'm fixing the WebDriver test, but the last Puppeteer test just won't play nice.
 
> @pmk65 What exactly does Pupeteer? Does it control your local browser?

>Puppeteer is a Node library which provides a high-level API to control headless Chrome or Chromium over the DevTools Protocol. It can also be configured to use full (non-headless) Chrome or Chromium. 

In CodeceptJS you use it as a helper, similar to WebDriver.

pmk65 commented:
Had to make a custom `I.pressKey` function, so that Puppeteer can accept strings instead of single characters. This should solve the string issue with WebDriver.

The Puppeteer popup problem in the array tests can be solved by rewriting the test, so that there's no double popups. (The value could be stored in a "result" textarea, similar to what is done in other tests.)

pmk65 commented:
Almost there. Now only fails in Firefox at:
> Given css selector expression "[name="root[integer]"]:read-only" is invalid: SyntaxError: '[name="root[integer]"]:read-only' is not a valid selector

Which is because in Firefox you will have to use `:-moz-read-only` and not `:read-only` 😠 
Guess I'll have to use a custom helper function to check for `disabled` and `readonly` states to ensure it works in all browsers/helpers .

It could either be a generic `I.grabBooleanAttributeFrom` or two separate helper functions `I.amDisabled` and `I.amReadOnly`. (Personally I prefer last solution) 

Example usage: 

`assert.equal(await I.grabBoooleanAttributeFrom('[name="root[checkbox]"]', 'disabled'), true);`
or
`assert.equal(await I.amDisabled('[name="root[checkbox]"]'), true);`

schmunk42 commented:
> Which is because in Firefox you will have to use `:-moz-read-only` and not `:read-only` 
> Guess I'll have to use a custom helper function to check for `disabled` and `readonly` states to ensure it works in all browsers/helpers .

Yeah, look like `-moz` is required for those.

schmunk42 commented on config/puppeteer.helpers.js:
```diff
@@ -0,0 +1,86 @@
+
+  // Custom myFunc function.
+  // Test function. Just calls click function
+  async myFunc(xpath) {
```
Can be removed

pmk65 replied:
Is removed in next commit.

schmunk42 commented on tests/codeceptjs/editors/rating_test.js:
```diff
@@ -4,9 +4,22 @@ Feature('rating');
+  //I.seeElement('[id="root[rating]4"]:disabled');
+  //I.seeElement('[id="root[rating]3"]:disabled');
+  //I.seeElement('[id="root[rating]2"]:disabled');
+  //I.seeElement('[id="root[rating]1"]:disabled');
```
Comments should be removed, applies to all commented code

pmk65 replied:
I just kept them as reference while testing since I had to commit new test changes to see if the tests passed on WebDriver/Selenium. I'll cleanup everything in the final commit.

I have added some extra helpers for testing if an element is disabled or readonly.
 * **I.amDisabled(xpath)**
 * **I.amNotDisabled(xpath)**
 * **I.amReadOnly(xpath)**
 * **I.amNotReadOnly(xpath)**

If any of these functions fail, the session will terminate with an error.

#### Functions that extends default functions
 * **I.pressKey(key)** - Extended so Puppeteer will press sequence of character if "Unknown key" like WebDriver
* **I.amAcceptingPopups()** - Extended so WebDriver will accept the Puppeteer command
* **I.amCancellingPopups()** - Extended so WebDriver will accept the Puppeteer command

#### Misc (Currently not used but might be useful later.)
* **grabBooleanAttributeFrom(xpath, attrib)** - Get real Boolean value from attribbute.
* **isTrue(val)** - Returns real Boolean value

pmk65 commented:
All test's now pass on Webdriver/Selenium, both Chrome and Firefox.

But don't merge yet. The array test still fails using Puppeteer due to the double popups. I will rewrite it so it works on both Puppeteer and WebDriver/Selenium.

schmunk42 commented:
> @schmunk42 The "advanced_test.js" test looks a bit odd. It ends with clicking on a button with className "json-editor-btn-delete" (Which all delete buttons have) without any additional checks or closing the popup dialog that opens up. Is this intentional?

I don't think so.

The changes LGTM overall. Is it ready from your side?

pmk65 commented:
> 
> 
> > @schmunk42 The "advanced_test.js" test looks a bit odd. It ends with clicking on a button with className "json-editor-btn-delete" (Which all delete buttons have) without any additional checks or closing the popup dialog that opens up. Is this intentional?
> 
> I don't think so.
> 
> The changes LGTM overall. Is it ready from your side?

Yes all is good to go,

schmunk42 commented:
Merged, thank you!

@pmk65 Since codeceptJS is now part of the `packages.json` we might be able to remove the Docker image, right?

pmk65 commented:
> 
> 
> Merged, thank you!
> 
> @pmk65 Since codeceptJS is now part of the `packages.json` we might be able to remove the Docker image, right?

I have no idea, since I can't test anything using Docker myself. So you'll have to try it out yourself.
